pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeRollupTaskGenerator.java

```diff
@@ -0,0 +1,491 @@
+
+        if (!isValidMergeWindowEndTime(windowEndMs, bufferMs, lowerMergeLevel, mergeRollupTaskMetadata)) {
+          LOGGER.info(
+              "Window with start: {} and end: {} of mergeLevel: {} is not a valid merge window, Skipping task generation: {}",
+              windowStartMs, windowEndMs, mergeLevel, taskType);
+          continue;
+        }
+
+        // Find all segments overlapping with the merge window, if all overlapping segments are merged, bump up the target window
+        List<OfflineSegmentZKMetadata> selectedSegments = new ArrayList<>();
+        boolean hasUnMergedSegments = false;
+        boolean isValidMergeWindow = true;
+
+        // The for loop terminates in following cases:
+        // 1. Found unmerged segments in target merge window, need to bump up watermark if windowStartMs > watermarkMs,
+        //    will schedule tasks
+        // 2. All segments are merged in the merge window and we have loop through all segments, need to bump up
+        //    watermark to windowStartMs + bucketTimeMs, skip scheduling
+        // 3. Current watermark > endTimeMs of all segments, won't bump up watermark, skip scheduling
+        // 4. Merge window is invalid (windowEndMs > System.currentTimeMillis() - bufferMs || windowEndMs > waterMark of
+        //    the lower mergeLevel), may need to bump up watermark if windowStartMs > watermarkMs, skip scheduling
+        for (OfflineSegmentZKMetadata preSelectedSegment : preSelectedSegments) {
+          if (windowStartMs <= preSelectedSegment.getEndTimeMs() && preSelectedSegment.getStartTimeMs() < windowEndMs) {
+            // For segments overlapping with current window, add to the result list
+            selectedSegments.add(preSelectedSegment);
+            if (!isMergedSegment(preSelectedSegment, mergeLevel)) {
+              hasUnMergedSegments = true;
+            }
+          } else if (windowEndMs <= preSelectedSegment.getStartTimeMs()) {
+            // Has gone through all overlapping segments for current window
+            if (hasUnMergedSegments) {
+              // Found unmerged segments, schedule merge task for current window
+              break;
+            } else {
+              // No unmerged segments found, clean up selected segments and bump up the merge window
+              // TODO: If there are many small merged segments, we should merge them again
+              selectedSegments = new ArrayList<>();
+              selectedSegments.add(preSelectedSegment);
+              if (!isMergedSegment(preSelectedSegment, mergeLevel)) {
+                hasUnMergedSegments = true;
+              }
+              while (windowEndMs <= preSelectedSegment.getStartTimeMs()) {
+                LOGGER.debug("No unmerged segments found for window [{}, {}), Update window to: [{}. {})",
+                    windowStartMs, windowEndMs, windowStartMs + bucketMs, windowEndMs + bucketMs);
+                windowStartMs += bucketMs;
```

Comment:
Directly calculate the `windowStartMs` by `preSelectedSegment.getStartTimeMs() / bucketMs * bucketMs` and avoid the unnecessary while loop

Reply:
Good point, updated.